File: SockStoreJava/src/application/usecases/src/main/java/be/itenium/sockstore/usecases/UpdateProductUseCase.java
package be.itenium.sockstore.usecases;

import be.itenium.sockstore.api.UpdateProduct;
import be.itenium.sockstore.api.commands.UpdateProductCommand;
import be.itenium.sockstore.domain.ProductAggregate;
import be.itenium.sockstore.domain.ProductPort;
import be.itenium.sockstore.vocabulary.Product;

public class UpdateProductUseCase implements UpdateProduct {
    private final ProductPort productPort;

    public UpdateProductUseCase(ProductPort productPort) {
        this.productPort = productPort;
    }

    public Product update(UpdateProductCommand command) {
        ProductAggregate updated = new ProductAggregate(
                command.id(),
                command.naam(),
                command.categorie(),
                command.prijs(),
                command.voorraad()
        );
        return productPort.save(updated).toProduct();
    }
}

File: SockstoreNet/Application/UseCases/UpdateProductUseCase.cs
using Application.Commands;
using Application.Interfaces;
using Application.Ports;
using Vocabulary;

namespace Application.UseCases;

public class UpdateProductUseCase(IProductPort productPort) : IUpdateProduct
{
    public async Task<Product> Update(UpdateProductCommand command)
    {
        var product = await productPort.FindById(command.Id, CancellationToken.None);
        if (product == null)
            throw new ArgumentException($"Product with id {command.Id} was not found");

        product.UpdateProduct(command.Naam, command.Categorie, command.Prijs);
        await productPort.Save(product);
        return product.ToProduct();
    }
}
